File: app/src/main/java/com/bubelov/coins/ui/fragment/SignInFragment.java
package com.bubelov.coins.ui.fragment;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.bubelov.coins.R;
import com.bubelov.coins.dagger.Injector;
import com.bubelov.coins.repository.user.UserRepository;
import com.bubelov.coins.ui.activity.AbstractActivity;
import com.bubelov.coins.ui.activity.MapActivity;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

/**
 * @author Igor Bubelov
 */

public class SignInFragment extends Fragment implements TextView.OnEditorActionListener {
    @BindView(R.id.email)
    EditText email;

    @BindView(R.id.password)
    EditText password;

    @BindView(R.id.sign_in)
    Button signInButton;

    @Inject
    UserRepository userRepository;

    private Unbinder unbinder;

    private SignInTask signInTask;

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        Injector.INSTANCE.getMainComponent().inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_sign_in, container, false);
        unbinder = ButterKnife.bind(this, view);
        password.setOnEditorActionListener(this);
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        password.setOnEditorActionListener(null);
        unbinder.unbind();

        if (signInTask != null) {
            signInTask.cancel(true);
        }
    }

    @Override
    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
        if (actionId == EditorInfo.IME_ACTION_GO) {
            signIn(email.getText().toString(), password.getText().toString());
            return true;
        }

        return false;
    }

    @OnClick(R.id.sign_in)
    void onSignInClick() {
        signIn(email.getText().toString(), password.getText().toString());
    }

    private void signIn(String email, String password) {
        signInTask = new SignInTask(email, password);
        signInTask.execute();
    }

    private class SignInTask extends AsyncTask<Void, Void, Boolean> {
        private String email;

        private String password;

        SignInTask(String email, String password) {
            this.email = email;
            this.password = password;
        }

        @Override
        protected void onPreExecute() {
            signInButton.setEnabled(false);
        }

        @Override
        protected Boolean doInBackground(Void... params) {
            return userRepository.signIn(email, password);
        }

        @Override
        protected void onPostExecute(Boolean success) {
            signInButton.setEnabled(true);

            if (!success) { // TODO
                if (getActivity() != null) {
                    AbstractActivity abstractActivity = (AbstractActivity) getActivity();
                    abstractActivity.showAlert(R.string.could_not_connect_to_server);
                }

                return;
            }

            Intent intent = new Intent(getContext(), MapActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(intent);
        }
    }
}

File: app/src/main/java/com/bubelov/coins/ui/fragment/SignInFragment.kt
/*
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

package com.bubelov.coins.ui.fragment

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.TextView

import com.bubelov.coins.R
import com.bubelov.coins.repository.user.SignInResult
import com.bubelov.coins.repository.user.UserRepository
import com.bubelov.coins.ui.activity.MapActivity
import dagger.android.support.AndroidSupportInjection

import javax.inject.Inject

import kotlinx.android.synthetic.main.fragment_sign_in.*
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch
import org.jetbrains.anko.alert

class SignInFragment : Fragment(), TextView.OnEditorActionListener {
    @Inject internal lateinit var userRepository: UserRepository

    override fun onAttach(context: Context) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_sign_in, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        password.setOnEditorActionListener(this)
        sign_in.setOnClickListener { signInNew(email.text.toString(), password.text.toString()) }
    }

    override fun onEditorAction(v: TextView, actionId: Int, event: KeyEvent?): Boolean {
        if (actionId == EditorInfo.IME_ACTION_GO) {
            signInNew(email.text.toString(), password.text.toString())
            return true
        }

        return false
    }

    private fun signInNew(email: String, password: String) = launch(UI) {
        sign_in_panel.visibility = View.GONE
        progress.visibility = View.VISIBLE

        val signInResult = async { userRepository.signIn(email, password) }.await()

        when (signInResult) {
            is SignInResult.Success -> startActivity(Intent(activity, MapActivity::class.java).apply { addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP) })
            is SignInResult.Error -> activity?.alert { message = signInResult.e.message ?: getString(R.string.something_went_wrong) }?.show()
        }

        sign_in_panel.visibility = View.VISIBLE
        progress.visibility = View.GONE
    }
}